feat(curriculum): add otp generator

zairahira opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes https://github.com/freeCodeCamp/CurriculumExpansion/issues/663

<!-- Feel free to add any additional description of changes below this line -->

zairahira committed:
define project structure

zairahira committed:
add user stories and some tests

zairahira committed:
Merge remote-tracking branch 'upstream/main' into feat/add-react-otp-generator

zairahira committed:
almost there

zairahira committed:
add disable button feature

zairahira committed:
Merge branch 'main' into feat/add-react-otp-generator

zairahira committed:
fix failing ci tests

zairahira committed:
add css for solution

naomi-lgbt commented on curriculum/challenges/english/25-front-end-development/lab-otp-generator/67c562286b29447da020d407.md:
```diff
@@ -0,0 +1,329 @@
+    - A `button` element with the text `"Generate OTP"`. When clicked, it should generate a new OTP and start a 5 -second countdown.
+    - The `"Generate OTP"` button should be disabled while the countdown timer is running.
+
+4. You should display an the following message when the countdown timer reaches `0`: `OTP expired. Click the button to generate a new OTP.`.
```
```suggestion
4. You should display the following message when the countdown timer reaches `0`: `OTP expired. Click the button to generate a new OTP.`.
```

zairahira committed:
fix typo

Co-authored-by: Naomi <accounts+github@nhcarrigan.com>

Ksound22 reviewed (commented):
This looks good. The only thing that confuses me is that we are not providing campers with the CSS.

Is that not how it's supposed to be?

zairahira commented:
> I also noticed we can't do something like this:
> 
> ```js
> import { useState, useEffect } from 'react';
> ```

I tried to add that line in the `jsx` file, but I see this error in the console:

```md
Uncaught TypeError: Cannot read properties of undefined (reading 'useState')
Uncaught TypeError: Cannot read properties of undefined (reading 'useState')
{ [TypeError: Cannot read properties of undefined (reading 'useState')] _suppressLogging: true }
The above error occurred in the <OTPGenerator> component:

    at OTPGenerator (about:srcdoc:87:32)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
Build error, open your browser console to learn more.
```

zairahira committed:
provide import statement in seed

zairahira committed:
replace placeholder

zairahira committed:
rename folders and dashed names

Ksound22 reviewed (approved):
LGTM 🚀

moT01 reviewed (changes requested):
This is a fun lab @zairahira 👍 Here's some findings that I think need fixing...

<hr>

The main problem I see is that the setTimeouts in the tests don't seem to be working.

Test 6 passes before I write any code, so the asserts aren't running there?
Other tests with timeouts pass when I don't have the correct code.

I think if you want the tests to wait, you can [add something similar to the tests here](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md#--hints--) - e.g:
```
await new Promise((resolve) => setTimeout(resolve, 100)); // Waits for 100ms
```

I would suggest to remove the timeout's where you can. They don't seem to be needed on some of the tests - but I'm not sure why you added them, so maybe they are. There is a few tests where they may be needed since we want to wait and make sure the things change after X seconds or whatever. Hey @ojeytonwilliams, you mentioned not using timeout's in the tests, any input here?

<hr>

Additional things...

- Many of the tests are checking for `.container p`, `.container h1`, etc. So if I want to add any extra elements in my container to give my project some flair - e.g. another `p`, `h1`, `h2`, etc - the tests my fail because they're checking my new element instead of the one with the correct info. Wonder if we should give the required elements an ID for the tests to check?

- Pretty minor, but I would prefer to use the full text for the title and dashed names of the challenge - e.g: `Build a One-Time Password Generator` and `build-a-one-time-password-generator`. I don't really mind leaving it either, but that's my preference. If we change it, probably change all the folders and dashed names everywhere. Any thoughts on that @zairahira @jdwilkin4 ?

- The ``1. You should use the `useState` hook.`` test is passing before I write any code. It looks like it's checking for the `useState` import via regex, but it doesn't check if it's actually being used. On that note, the user stories say to use the `useEffect` hook, not the `useState` hook - so we should probably check for that instead? I would change that to `useEffect` and then find a way to test that it's actually being used. You could add an additional user story and test about using `useState` as well if you want, since that will likely be needed - but I don't think those are necessary - up to you.

- User story 4 - the test seems to be checking the `p` element, but the user story or test don't mention that. I might move this user story up to the `p` part of user story 3. The `p` has three states: empty, `"Expires in: X seconds"`, and `"OTP expired. Click the button to generate a new OTP."` So I would keep them all together to make it more clear. Also, probably mention the `p` in the test for that one.

<details><summary>Here's the code I used to pass:</summary>

```
const { useState, useEffect } = React;

export const OTPGenerator = () => {
  const [otp, setOtp] = useState('');
  const [timeLeft, setTimeLeft] = useState(0);

  useEffect(() => {
    if (timeLeft > 0) {
      const timer = setInterval(() => {
        setTimeLeft((prev) => prev - 1);
      }, 1000);
      return () => clearInterval(timer);
    }
  }, [timeLeft]);

  const generateOtp = () => {
    setOtp(Math.floor(100000 + Math.random() * 900000));
    setTimeLeft(5);
  }

  return (
    <div className="container">
      <h1>OTP Generator</h1>
      <h2>{timeLeft > 0 ? otp : "Click 'Generate OTP' to get a code"}</h2>
      <p>{timeLeft > 0 ? `Expires in ${timeLeft} seconds` : otp === '' ? '' : 'OTP Expired'}</p>
      <button onClick={generateOtp} disabled={timeLeft > 0}>Generate OTP</button>
    </div>
  );
};
```

</details>

^ note that the code passes but doesn't fulfill the user stories because some of the tests aren't working quite right.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

zairahira committed:
update dashed names

zairahira committed:
update title

zairahira committed:
update user story 3

zairahira committed:
add ids and tests for container elements

zairahira committed:
add a few more tests

moT01 committed:
Merge branch 'main' into feat/add-react-otp-generator

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-one-time-password-generator/67c562286b29447da020d407.md:
````diff
@@ -0,0 +1,341 @@
+assert.strictEqual(document.querySelector('.container h2#otp-display').textContent.trim(), "Click 'Generate OTP' to get a code");
+```
+
+The `.container` should contain a `p` element with the ID `otp-timer` that is initially empty.
````
```suggestion
Initially, the `.container` should not contain any `p` elements.
```
The solution starts without them and only shows one after the button is clicked.

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-one-time-password-generator/67c562286b29447da020d407.md:
````diff
@@ -0,0 +1,341 @@
+The `.container` should contain a `p` element with the ID `otp-timer` that is initially empty.
+
+```js
+setTimeout(() => {
+  const paragraph = document.querySelector('.container p#otp-timer');
+  assert.exists(paragraph);
+  assert.equal(paragraph.textContent.trim(), "", "The paragraph element should be empty initially");
+}, 100);
````
```suggestion
const paragraph = document.querySelector('.container p');
assert.notExists(paragraph);
```

ojeytonwilliams commented:
Hey @zairahira, I found some issues with the solution while I was testing out the fake timers.

As it stands, the `useEffect` spawns a new interval and cancels the old one each time the dependencies change. This, effectively, turns `setInterval` into `setTimeout`.

The real problem is that this is not how `useEffect` should be used. It's not intended to trigger things, it's another way to respond to changing state or to sync with external systems. In this case if we call `setInterval` inside `handleGenerateOtp`, then issue goes away and the code is cleaner.

Here's how I think we should handle it:

```jsx
const { useState, useEffect, useRef } = React;

export const OTPGenerator = () => {
  const intervalId = useRef(null);
  const [otp, setOtp] = useState(null);
  const [secondsLeft, setSecondsLeft] = useState(5);
  const [isCounting, setIsCounting] = useState(false);
  const [hasGeneratedOtp, setHasGeneratedOtp] = useState(false);

  const generateOTP = () => Math.floor(100000 + Math.random() * 900000);

  const handleGenerateOtp = () => {
    setOtp(generateOTP());
    setSecondsLeft(5);
    setIsCounting(true);
    setHasGeneratedOtp(true);
    intervalId.current = setInterval(() => {
      setSecondsLeft(prevSeconds => prevSeconds - 1);
    }, 1000);
  };

  useEffect(() => {
    if (secondsLeft === 0) {
      setIsCounting(false);
      clearInterval(intervalId.current);
    }
  }, [secondsLeft]);

  useEffect(
    () => () => {
      clearInterval(intervalId.current);
    },
    []
  );

  return (
    <div className='container'>
      <h1 id='otp-title'>OTP Generator</h1>
      <h2 id='otp-display'>
        {otp ? otp : "Click 'Generate OTP' to get a code"}
      </h2>
      {isCounting ? (
        <p id='otp-timer'>Expires in: {secondsLeft} seconds</p>
      ) : (
        hasGeneratedOtp && (
          <p id='otp-expired-message'>
            OTP expired. Click the button to generate a new OTP.
          </p>
        )
      )}
      <button
        id='generate-otp-button'
        onClick={handleGenerateOtp}
        disabled={isCounting}
      >
        Generate OTP
      </button>
    </div>
  );
};
```

It's still necessary to `useEffect`s to stop the interval, but it's now clear what their jobs are: stopping the countdown and cleaning up, respectively.

zairahira committed:
reword for crowdin

moT01 committed:
Merge branch 'main' into feat/add-react-otp-generator

zairahira committed:
apply suggestions

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-one-time-password-generator/67c562286b29447da020d407.md:
```diff
@@ -0,0 +1,346 @@
+<head>
+    <meta charset="UTF-8" />
+    <title>OTP Generator</title>
+    <script src="https://unpkg.com/react@18/umd/react.development.js"></script>
+    <script src="https://unpkg.com/react-dom@18/umd/react-dom.development.js"></script>
+    <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
```
```suggestion
    <script src="https://cdnjs.cloudflare.com/ajax/libs/react/18.3.1/umd/react.development.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/react-dom/18.3.1/umd/react-dom.development.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-standalone/7.26.5/babel.min.js"></script>
```
We're using cdnjs since unpkg suffered a few outages: https://github.com/freeCodeCamp/freeCodeCamp/pull/59291

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-one-time-password-generator/67c562286b29447da020d407.md:
```diff
@@ -0,0 +1,346 @@
+    <meta charset="UTF-8" />
+    <title>Color Picker</title>
+    <link rel="stylesheet" href="styles.css" />
+    <script src="https://unpkg.com/react@18/umd/react.development.js"></script>
+    <script src="https://unpkg.com/react-dom@18/umd/react-dom.development.js"></script>
+    <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
```
```suggestion
    <script src="https://cdnjs.cloudflare.com/ajax/libs/react/18.3.1/umd/react.development.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/react-dom/18.3.1/umd/react-dom.development.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-standalone/7.26.5/babel.min.js"></script>
```
See below

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-one-time-password-generator/67c562286b29447da020d407.md:
````diff
@@ -0,0 +1,346 @@
+You should export the `OTPGenerator` component.
+
+```js
+assert.isFunction(window.index.OTPGenerator, 1);
````
```suggestion
assert.isFunction(window.index.OTPGenerator);
```
It's not clear what the message is for.

ojeytonwilliams reviewed (changes requested):
Hi Zaira, thanks for making those updates.

While taking another look I noticed that the setTimeouts used in the tests all resolve after the test has passed or failed and can't change the outcome.

The easiest make the tests wait for the timeouts is to wrap the test in an async function and create promises for the timeouts. e.g.

```js
async () => {
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

await delay(10)
assert(...)
}
```

More generally, I would recommend writing each test before adding the code that makes it pass. That way you can be confident that the test really is checking what you want it to.

zairahira committed:
Apply suggestions from code review

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

zairahira commented:
Hey @ojeytonwilliams,

I have tried writing one of the tests using the suggested format but it is not working:

````md

```js
async () => {
  const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

  const button = document.querySelector(".container button#generate-otp-button");
  const h2 = document.querySelector(".container h2#otp-display");

  button.dispatchEvent(new Event('click'));
  await delay(300);

  const p = document.querySelector(".container p#otp-timer");

  assert.match(h2.textContent, /^[0-9]{6}$/, "OTP should be a 6-digit number");
  assert.match(p.textContent, /Expires in: 5 seconds/, "Countdown should start at 5 seconds");
```

````

ojeytonwilliams commented:
Hi Zaira, if you use `button.cllck();` rather than `button.dispatchEvent(new Event('click'));` it should work.

zairahira committed:
midify the use of async

ojeytonwilliams commented:
If I understand the goal correctly, you want a test that looks for a `.container` element and checks that element doesn't have any `p` children. How about this?

````md

Initially, there should be a `.container` element that does not contain any `<p>` elements.

```js
const container = document.querySelector('.container');
assert.exists(container, "The container element should exist");
const pElements = document.querySelectorAll('.container p');
assert.strictEqual(pElements.length, 0, "Initially, there should be no <p> elements inside '.container'");
```
````

ojeytonwilliams commented:
Also, now the tests are running, they're taking longer than the timeout. I've got fake timers working in my test runner and it's very nearly done. However, if we need to get this in ASAP, we'll have to increase the timeouts.

Do you know when this needs to be live?

zairahira commented:
We are trying to launch this by the end of this month. But if more time is needed to make the fake timers work, we can increase the timelines.

zairahira committed:
update test 6

ojeytonwilliams commented:
The end of the month should be doable. I should be done next week and then it will need reviewing.

moT01 commented on curriculum/challenges/_meta/lab-one-time-password-generator/meta.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "Build a One-Time Password Generator",
+  "isUpcomingChange": false,
```
```suggestion
  "isUpcomingChange": true,
```

moT01 committed:
Update curriculum/challenges/_meta/lab-one-time-password-generator/meta.json

moT01 committed:
Merge branch 'main' into feat/add-react-otp-generator

ojeytonwilliams commented:
I'll fix the conflicts and add the test updates.

ojeytonwilliams committed:
Merge branch 'main' into feat/add-react-otp-generator

ojeytonwilliams committed:
feat: release otp lab

ojeytonwilliams committed:
fix: update tests to use fake timers

I also updated the solution so it passes the new tests.

ojeytonwilliams committed:
fix: remove unnecessary runAllAsync

I'm not sure what was waiting for, but the React component has finished
already finished making all the updates we're looking at, so it's not
necessary.

ojeytonwilliams committed:
feat: clean up fake timers

ojeytonwilliams commented:
Hey @zairahira @moT01 I've added one final test to uninstall the clock so that `setTimeout` functions normally in other lessons. Obviously this is a dirty hack and it shows a test result to the user (which sucks), but we need something like this.

When I get a chance, I'll implement `afterAll` for this kind of thing, but that's the next major release of the test runner. Coming soon, but not this week.
